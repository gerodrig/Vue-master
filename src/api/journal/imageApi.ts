import { v2 as cloudinary } from 'cloudinary';

cloudinary.config({
    cloud_name: import.meta.env.VITE_CLOUDINARY_CLOUD_NAME,
    api_key: import.meta.env.VITE_CLOUDINARY_API_KEY,
    api_secret: import.meta.env.VITE_CLOUDINARY_API_SECRET,
    secure: true,
  });

export const deleteImage = async (url: string | null): Promise<boolean> => {

    if(!url) return false;

    if (url) {
        //url format https://res.cloudinary.com/gerodrig18/image/upload/v1682391219/vue-test/wbk4pifuthuyxqn08ly1.jpg

        // get vue-test/wbk4pifuthuyxqn08ly1
        // const public_id = url.split('/').pop()?.split('.')[0] || '';
        const public_id = url.split('/').slice(-2).join('/').split('.')[0] || '';
  
        try {
          const result = await cloudinary.api.delete_resources([public_id], {});
            //check if deleted: {'deleted': {'vue-test/': 'deleted'}} includes deleted return true
            //else return false
            if(Object.values(result.deleted)[0] === 'deleted') return true;
            
        } catch (error) {
          console.log(error);
        }
      }

      return false;
};