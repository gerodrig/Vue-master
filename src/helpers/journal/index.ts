export {default as uploadImage} from './uploadImage';
export { getDate } from "./getDate";


