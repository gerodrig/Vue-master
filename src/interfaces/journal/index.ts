export interface JournalState {
        isLoading: boolean;
        entries: Entry[];
}

export interface AuthState {
    status: AuthStatus;
    user: null | User;
    idToken: null | string;
    refreshToken: null | string;
}

export interface User {
    username?: string;
    email: string;
    password?: string;
}


//enum for the auth status
export type AuthStatus = 'authenticated' | 'not-authenticated' | 'authenticating';

export interface Rootstate {
    journal?: JournalState;
    auth?: AuthState;
}

export interface Entry {
    id: string;
    date: string | number;
    text: string;
    picture?: string | null;
}

export interface FormattedDate {
    day: number;
    month: string;
    year: number;
    dayOfWeek: string;
  }

export interface EntryFirebase {
    [ key: string ]: {
        date: string;
        text: string;
        picture: string | null;
    }
}