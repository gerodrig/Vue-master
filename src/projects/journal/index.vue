<template>
  <!-- <AuthPage /> -->
  <router-view />
  <!-- <div class="flex flex-col items-center justify-center h-screen text-3xl">
    <Logo />
    <h1 class="mb-4 text-5xl">Journal App</h1>
    <button
      class="px-4 py-2 mt-4 font-bold text-white bg-green-700 rounded hover:bg-green-800"
    >
      <RouterLink :to="{ name: 'no-entry' }">Click to launch App</RouterLink>
    </button>
  </div> -->
</template>

<script lang="ts">
import useAuth from '@/composables/auth/useAuth';
import { defineAsyncComponent, watch } from 'vue';
import { useRouter } from 'vue-router';

export default {
  name: 'Journal App',
  components: {
    Logo: defineAsyncComponent(() => import('@/components/shared/Logo.vue')),
    AuthPage: defineAsyncComponent(
      () => import('@/auth/journal/layouts/AuthLayout.vue')
    ),
  },
  setup() {
    const { checkAuthStatus, authStatus } = useAuth();
    const router = useRouter();

    watch(authStatus, () => {
      if(authStatus.value !== 'authenticated') {
        router.push({ name: 'login' });
      }
    });

    checkAuthStatus();
    return {
      authStatus,
    };
  },
};
</script>
