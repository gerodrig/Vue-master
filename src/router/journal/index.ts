import { RouteLocationNormalized as Route } from 'vue-router';

export const RouteNames = {
    HOME: 'daybook',
    ENTRY: 'entry',
    NOENTRY: 'no-entry',
  }

export default {
    name: RouteNames.HOME,
    component: () => import(/* webpackChunkName: "daybook" */ '@layouts/journal/DayBookLayout.vue'), 
    children: [
        {
            path: '',
            name: RouteNames.NOENTRY,
            component: () => import(/* webpackChunkName: "daybook-no-entry" */ '@/views/journal/NoEntrySelected.vue') 

        },
        {
            path: ':id',
            name: RouteNames.ENTRY,
            component: () => import(/* webpackChunkName: "daybook-entry" */ '@/views/journal/EntryView.vue'),
            props: (route: Route) => {
                return {
                    id: route.params.id
                }
            }
        }
    ],
}

