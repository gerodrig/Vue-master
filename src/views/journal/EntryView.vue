<template>
  <template v-if="entry">
    <div class="flex justify-between p-2">
      <div>
        <span class="text-3xl font-bold text-green-600">{{ day }}</span>
        <span class="mx-1 text-2xl">{{ month }}</span>
        <span class="mx-2 text-xl font-light">{{ year }}, {{ dayOfWeek }}</span>
      </div>
      <div>
        <input
          type="file"
          @change="onSelectedImage"
          v-show="false"
          accept="image/*"
          ref="imageSelector"
        />
        <button
          v-if="entry.id && entry.id !== 'new'"
          class="px-4 py-2 mx-2 font-bold text-white bg-red-500 rounded hover:bg-red-600"
          @click="deleteEntry"
        >
          Delete
          <i class="fa fa-trash-alt"></i>
        </button>
        <button
          class="px-4 py-2 font-bold text-white bg-blue-700 rounded hover:bg-blue-900"
          @click="onUploadImage"
        >
          Upload photo
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>
    <hr />
    <div class="flex flex-col h-screen px-3">
      <textarea
        class="flex-1 h-full p-2 mt-4 text-xl border-2 outline-emerald-700 bg-green-50"
        placeholder="What happened today?"
        v-model="entry.text"
      ></textarea>
    </div>
    <img
      v-if="entry.picture && !loadedImage"
      :src="entry.picture"
      alt="entry-picture"
      class="fixed w-48 border border-gray-400 rounded-md shadow-md bottom-36 right-5"
    />
    <img
      v-if="loadedImage"
      :src="loadedImage"
      alt="entry-picture"
      class="fixed w-48 border border-gray-400 rounded-md shadow-md bottom-36 right-5"
    />
    <Fab icon="fa-save" @on:click="saveEntry" />
  </template>
</template>

<script lang="ts">
import { defineAsyncComponent } from "vue";
import useEntries from "@/composables/journal/useEntries";

export default {
  name: "EntryView",
  props: {
    id: {
      type: String,
      required: true,
    },
  },
  components: {
    Fab: defineAsyncComponent(() => import("@/components/journal/Fab.vue")),
  },
  setup(props) {

    const {
      id,
      entry,
      day,
      month,
      year,
      dayOfWeek,
      loadedImage,
      imageSelector,
  
      //methods
      saveEntry,
      deleteEntry,
      onSelectedImage,
      onUploadImage,

    } = useEntries(props);

    return {
      id,
      entry,
      day,
      month,
      year,
      dayOfWeek,
      loadedImage,
      imageSelector,
  
      //methods
      saveEntry,
      deleteEntry,
      onSelectedImage,
      onUploadImage,
    };
},
};
</script>




