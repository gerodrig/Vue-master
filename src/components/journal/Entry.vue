<template>
  <div
    class="p-2 mb-3 transition-all duration-200 ease-in border-b border-green-600 cursor-pointer hover:bg-slate-300"
    @click="$router.push({ name: 'entry', params: { id } })"
  >
    <!-- Title -->
    <div class="flex">
      <span class="text-lg font-bold text-green-600">{{ day }}</span>
      <span class="mx-1 font-bold">{{ month }}</span>
      <span class="mx-2 font-light">{{ year }}, {{ dayOfWeek }}</span>
    </div>

    <div class="text-base">
      {{ text }}
    </div>
  </div>
</template>

<script lang="ts">
import { computed, toRef } from 'vue';
import { Entry, FormattedDate} from '@interfaces/journal/index';
import { getDate } from '../../helpers/journal/getDate';



export default {
  name: 'Entry',
  props: {
    entry: {
      type: Object as () => Entry,
      required: true,
    },
  },
  setup(props ) {
    const entry = toRef(props, 'entry');
    const date = computed<FormattedDate>(() => getDate(entry.value));

    const shortText = computed(() => {
      const text = entry.value.text;
      return text.length > 100 ? text.substring(0, 100) + '...' : text;
    });

    return {
      ...date.value,
      text: shortText,
      id: computed(() => entry.value.id),
    };
  },
};
</script>
