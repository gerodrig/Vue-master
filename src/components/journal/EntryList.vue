
<template>
    <div class="h-screen mx-4 mt-4 border-r border-green-900 entry-list-container">
        <input 
            type="text"
            class="items-center w-11/12 px-2 py-2 mx-auto rounded-md shadow-md ring-1 ring-gray-400 focus:border-green-500 focus:ring focus:ring-green-200 focus:ring-opacity-50"
            placeholder="Search Entry"
            v-model="term"
        />

        <div class="flex flex-col items-center justify-center my-4">
            <button @click="$router.push({name: 'entry', params: {id: 'new'}})" class="w-2/3 px-4 py-2 font-semibold text-white bg-green-800 rounded hover:bg-green-900 ">
                <i class="fas fa-plus-circle"></i>
                New Entry
            </button>
        </div>

        <div class="overflow-scroll entry-scroll-area">
            <Entry :entry="entry" v-for="entry in entries" :key="entry.id" />
        </div>
    </div>
</template>

<script lang="ts">
import { defineAsyncComponent, computed, ref, watch } from 'vue';
import { useStore } from 'vuex';
import { Rootstate, Entry } from '@interfaces/journal/index';

export default {
    name: 'EntryList',
    components: {
        Entry: defineAsyncComponent(() => import('@components/journal/Entry.vue'))
    },
    setup() {
        const store = useStore<Rootstate>();
        const term = ref('');
        const getEntriesByTerm = (value: string): Entry[] =>  store.getters['journal/getEntriesByTerm'](value);
        
        const entries =  computed<Entry[]>(() => {
            if (term.value.length > 2) {
                return getEntriesByTerm(term.value);
            } else {
                return store.state.journal.entries;
            }
        });
        
        return {
            term,
            entries
        }
    }
}
</script>

<style   scoped >
.entry-list-container {
    height: calc(100vh - 56px);
}

.entry-scroll-area {
    height: calc( 100vh - 120px);
}

</style>