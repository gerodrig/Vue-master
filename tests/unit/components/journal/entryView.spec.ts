import {beforeEach, describe, expect, it, vi } from 'vitest';
import { shallowMount, VueWrapper } from '@vue/test-utils';
import { createStore } from 'vuex';

import Swal from 'sweetalert2';

import journal from '../../mocks/journal/test-journal-state';
import EntryView from '@/views/journal/EntryView.vue';
import { useRouter } from 'vue-router';

vi.mock('vue-router', () => ({
    useRouter: vi.fn(() => ({
        push: vi.fn()
    }))
}));

// vi.mock('vue-router')

//create sweeat alert mock
vi.mock('sweetalert2');

describe('EntryView components test', () => {
    let wrapper: VueWrapper<any>; 
    const mockedSwal = Swal as any;
// configure the store with the journal module
    const journalMockModule = {
        namespaced: true,
        getters: {
            getEntriesById: vi.fn(() => () => journal.entries[0])
        },
        state: () => ({
            isLoading: false,
            entries: journal.entries
        })
    };

    const store = createStore({
        modules: {
            journal: {...journalMockModule}
        }
    });

    //control the store dispatch and mutations
    store.dispatch = vi.fn() as any;

    beforeEach(async () => {
        vi.clearAllMocks();
    
        wrapper = shallowMount(EntryView, {
            props: {
                id: journal.entries[0].id
            },
            global: {

            plugins: [store],
            },
        });
    });

  it('should redirect the user because id does not exists.', () => {

        const push = vi.fn();
        const mockedRouter = vi.mocked(useRouter as any).mockReturnValue({
            push
        });

        const wrapper = shallowMount(EntryView, {
        props: {
            id: 'Id that does not exists'
        },
        global: {
        stubs: ['router-link', 'router-view'],
        plugins: [store],
        },
        });

        expect(wrapper.vm.$props.id).toBeTruthy();
        expect(mockedRouter).toHaveBeenCalled();
    });

    it('should match the entry correctly snapshot', () => {
        const push = vi.fn();
        const mockedRouter = vi.mocked(useRouter as any).mockReturnValue({
            push
        });

        expect(wrapper.vm.$props.id).toBe(journal.entries[0].id)
        expect(wrapper.html()).toMatchSnapshot();
        expect(push).not.toHaveBeenCalled();
    }
    );

    it('should not delete entry if swal option is cancelled', async () => {
        
        mockedSwal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: false}));
        const deleteBtn = wrapper.find('.bg-red-500');
        await deleteBtn.trigger('click');

        expect(mockedSwal.fire).toHaveBeenCalled();
        expect(mockedSwal.fire).toHaveBeenCalledWith(
            {
                title: 'Are you sure?',
                text: "You won't be able to revert this!",
                icon: 'warning',
                showCancelButton: true,
                confirmButtonColor: '#4CAF50',
                cancelButtonColor: '#d33',
                confirmButtonText: 'Yes, delete it!'
              })
    
    });

    it('should delete entry if swal option is confirmed', async () => {

            const push = vi.fn();
            const mockedRouter = vi.mocked(useRouter as any).mockReturnValue({
                push
            });
            
            mockedSwal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: true}));

            const deleteBtn = wrapper.find('.bg-red-500');
            await deleteBtn.trigger('click');
            // console.log(mockedSwal.fire.mock.calls);
    
            expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', journal.entries[0]);
            expect(mockedRouter).toHaveBeenCalled();
    });

});