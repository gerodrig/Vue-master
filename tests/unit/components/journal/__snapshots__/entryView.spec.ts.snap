// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EntryView components test > should match the entry correctly snapshot 1`] = `
"<div class=\\"flex justify-between p-2\\">
  <div><span class=\\"text-3xl font-bold text-green-600\\">18</span><span class=\\"mx-1 text-2xl\\">February</span><span class=\\"mx-2 text-xl font-light\\">2022, Friday</span></div>
  <div><input type=\\"file\\" accept=\\"image/*\\" style=\\"display: none;\\"><button class=\\"px-4 py-2 mx-2 font-bold text-white bg-red-500 rounded hover:bg-red-600\\"> Delete <i class=\\"fa fa-trash-alt\\"></i></button><button class=\\"px-4 py-2 font-bold text-white bg-blue-700 rounded hover:bg-blue-900\\"> Upload photo <i class=\\"fa fa-upload\\"></i></button></div>
</div>
<hr>
<div class=\\"flex flex-col h-screen px-3\\"><textarea class=\\"flex-1 h-full p-2 mt-4 text-xl border-2 outline-emerald-700 bg-green-50\\" placeholder=\\"What happened today?\\"></textarea></div>
<!--v-if-->
<!--v-if-->
<fab-stub icon=\\"fa-save\\"></fab-stub>"
`;
